### PR TITLE
basic integral

### PR DESCRIPTION
**Pull request by BowenFu**

### CODE REVIEW

File: racket-cas/racket-cas.rkt [hunk 1]
```diff
@@ -1998,6 +2001,183 @@
...
+      )
+  ))
+
+(module+ test
+  (check-equal? (try-subst-integrate (⊗ (Exp x) x) x '(expt @e x)) '(+ (* -1 (expt @e x)) (* (expt @e x) x)))
+  (check-equal? (try-subst-integrate (Sqr (⊕ x 1)) x '(+ x 1)) '(* 1/3 (expt (+ 1 x) 3)))
+  )
+
+(define (definite-integrate u x a b)
+  (define integral (integrate u x))
```
Comment (soegaard) [inline racket-cas/racket-cas.rkt]: ; symbolic computation of the definite integral of u wrt to x with limits a and b.
Is the limits a and b meant to be finite - or is ≠inf.0 allowed?

File: racket-cas/racket-cas.rkt [hunk 2]
```diff
@@ -1998,6 +2001,183 @@
...
+                '(+ (* 6 x) (* -2 (cos x)) (* -1/2 (expt (cos x) 2)) (* 3 (sin x))))
+  (check-equal? (integrate '(* (sin x) (expt (cos x) -4)) x) '(* 1/3 (expt (cos x) -3)))
+  (check-equal? (integrate '(* (expt (+ (sin x) 4) 3) (cos x)) x) '(* 1/4 (expt (+ 4 (sin x)) 4)))
+  (check-equal? (integrate '(* (+ x 2) (expt (+ (expt x 2) (* 4 x) 2) -1)) x)
+                '(* -1/2 (ln (expt (+ 2 (* 4 x) (expt x 2)) -1))))
+  ; Fix this with fraction simplification / divide gcd.
+  ; (check-equal? (integrate '(* (sin (+ (* a (expt x 2)) (* b (expt x 2)))) x) x) ...)
+  ; definite integral
+  (check-equal? (integrate `u x `a `b) '(+ (* -1 a u) (* b u)))
+  )
```
Comment (soegaard) [inline racket-cas/racket-cas.rkt]: This is already a nice range of expressions that can be integrated.

If you need inspiration, I can recommend looking at: Yacas
https://yacas.readthedocs.io/en/latest/book_of_algorithms/integration.html
The test suite is here:
https://github.com/grzegorzmazur/yacas/blob/master/tests/integrate.yts
And their rule set is here:
https://github.com/grzegorzmazur/yacas/blob/master/scripts/integrate.rep/code.ys

File: racket-cas/racket-cas.rkt [hunk 3]
```diff
@@ -2744,6 +2927,10 @@
           [(Equal (⊕ u w) v)     #:when (free-of w x) (r (Equal u (⊖ v w)))]
           [(Equal (⊗ w u) v)     #:when (free-of w x) (r (Equal u (⊘ v w)))]
           [(Equal (⊗ u w) v)     #:when (free-of w x) (r (Equal u (⊘ v w)))]
+          [(Equal (Ln v) w)      #:when (free-of w x)
+                                 (r (Equal v (Exp w)))]
+          [(Equal (Log u v) w)   #:when (and (free-of u x) (free-of w x))
+                                 (r (Equal v (Expt u w)))]
           [(Equal (Expt @e u) s) #:when (> s 0)        (r (Equal u (Ln s)))]
```
Comment (soegaard) [inline racket-cas/racket-cas.rkt]: Nice catch. Can you make a separate pull request with these improvements of solve?

File: racket-cas/racket-cas.rkt [hunk 4]
```diff
@@ -2851,7 +3040,8 @@
   (define (extract u)
     (match u
       [(list 'or e ...) (map solution e)]
-      [(list _ '= x0)   (list x0)]
+      [(list '= y x0) #:when (equal? y x)
+                      (list x0)]
       [_                '()]))
```
Comment (soegaard) [inline racket-cas/racket-cas.rkt]: Nice catch. Include this in the solve PR.
Is there a test that can catch these problems?

File: racket-cas/racket-cas.rkt [hunk 5]
```diff
@@ -1245,11 +1246,11 @@
...
-             [(and (= b 0) (= d 0)) `(expt ,u ,v)] ; u might be a product
+             [(and (equal? b 0) (equal? d 0)) `(expt ,u ,v)] ; u might be a product
              ; avoid computing the partition into real and imaginary multiple times,
              ; so pass the a, b c and along.
-             [(= d 0)               (ComplexRealExpt a b c)]
-             [(= b 0)               (RealComplexExpt a c d)]
+             [(equal? d 0)               (ComplexRealExpt a b c)]
+             [(equal? b 0)               (RealComplexExpt a c d)]
              [else                  (ComplexComplexExpt a b c d)])]
           [(_ _)          `(expt ,u ,v)])])]))
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply racket-cas/racket-cas.rkt]: I'll create a separate PR for it.

File: racket-cas/racket-cas.rkt [hunk 6]
```diff
@@ -1769,6 +1770,7 @@
     [(⊕ (⊗ p (Integer _) @pi) u) #:when (even? p) (Cos: u)]
     
     [(Acos u) u]    ; xxx only of -1<u<1
+    [(Asin u) (⊖ (⊗ 1/2 @pi) u)]
     [(⊖ u) (Cos u)] ; even function
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply racket-cas/racket-cas.rkt]: I'll create a separate PR for it.
Actually this simplification is needed to pass some `integrate` test cases.

File: racket-cas/racket-cas.rkt (hunk 1 above)
Comment (BowenFu) [reply racket-cas/racket-cas.rkt]: Only the basic cases are considered in the PR.

**Comment by BowenFu** [comment]
Thanks for your review. I have now added some doc for the functions. I hope I could have a better name for them.

File: racket-cas/racket-cas.rkt [hunk 7]
```diff
@@ -2868,6 +3230,74 @@
       [(⊖ u)  #f]
       [_      #t])) ; todo: add cases for variables with assumptions
 
+(define (solve-by-inverse-stacks w x)
```
Comment (BowenFu) [inline racket-cas/racket-cas.rkt]: Hi @soegaard , do you have any suggestions on how to share codes between `solve-by-inverse-stacks` and `solve-by-inverse`?

Two differences from `solve-by-inverse`:

1. `solve-by-inverse-stacks` needs to return the list of all intermediate results/eqns.
2. the `let/ec return`.

**Comment by BowenFu** [comment]
Hi @soegaard , the MR is almost stable now, can you help review it?

File: racket-cas/racket-cas.rkt (hunk 7 above)
Comment (BowenFu) [reply racket-cas/racket-cas.rkt]: THe function has been removed.

**Comment by soegaard** [comment]
Hi,

I just briefly skimmed the PR, but due to the many individual commits, the Github interface makes it hard to get the larger picture.

Can you make a new PR with a single commit - and use `git rebase` to make the commit on top of the current master?

One thing caught my eye though: you have made a matcher Sum to match `(list* '+ us)`.
Although natural I prefer to save the name Sum for a real sum operator like "sum from i=0 to 10 i^2" kind of thing. 

One option is to rename it to, say, HeadPlus (since the head of the form is a plus).
Another is to keep `(list* '+ us)` since this makes it obvious that the pattern matcher doesn't do any rewriting.

I also noted that the integrator can handle quite a large set of expressions - great!

**Comment by BowenFu** [comment]
> Hi,
> 
> I just briefly skimmed the PR, but due to the many individual commits, the Github interface makes it hard to get the larger picture.
> 
> Can you make a new PR with a single commit - and use `git rebase` to make the commit on top of the current master?
> 
> One thing caught my eye though: you have made a matcher Sum to match `(list* '+ us)`.
> Although natural I prefer to save the name Sum for a real sum operator like "sum from i=0 to 10 i^2" kind of thing.
> 
> One option is to rename it to, say, HeadPlus (since the head of the form is a plus).
> Another is to keep `(list* '+ us)` since this makes it obvious that the pattern matcher doesn't do any rewriting.
> 
> I also noted that the integrator can handle quite a large set of expressions - great!

Hi @soegaard , the overall changes can be seen in the `File changed` tab. I have rebased my commits now.
Sum / Prod is the orig matcher in the codes. I added two, SumTerms and ProdTerms, now I have renamed them to PlusTerms and TimesTerms.

**Comment by soegaard** [comment]
Thanks for rebasing. 

I missed that Sum/Prod was in the original code. Sorry about that.

I am working making racket-cas easier to modify by splitting racket-cas.rkt up in smaller modules.
I hope the change will improve compile time (and maybe make github commits easier to read?).
It's still a work in progress - see the "refactor" branch. It's a bit tricky to avoid circular module dependcies.
I'll keep you posted.

File: racket-cas/racket-cas.rkt [hunk 8]
```diff
@@ -828,6 +856,7 @@
     [(Expt (⊕ u v) n) #:when (and (>= n 3) (even? n))
                       (let ([t (e (Expt (⊕ u v) (/ n 2)))])
                         (e (⊗ t t)))]
+    [(Expt (⊕ u v) p-) (Expt (e (Expt (⊕ u v) (- p-))) -1)]
```
[the remark this answers is not in the document]
Comment (soegaard) [reply racket-cas/racket-cas.rkt]: What's the intended meaning of the pattern p- ?
Normally p is an integer, but p- also matches -42.5.
Perhaps p- ought to match negative integers only?

    `(check-equal? (math-match -42.5 [p+ #f] [p- #t] [_ #f]) #f)`

File: racket-cas/racket-cas.rkt [hunk 9]
```diff
@@ -918,11 +948,12 @@
 
 ; divide u by v
 (define (Oslash: u v)
-  (math-match* (u v)
-    [(r  0) +nan.0]
+  (match* (u v)
+    [((num: r)  0) +nan.0]
     [(u  1) u]
```
Comment (soegaard) [inline racket-cas/racket-cas.rkt]: Why change this?

**Comment by BowenFu** [comment]
> soegaard 3 hours ago  Owner
> What's the intended meaning of the pattern p- ?
> Normally p is an integer, but p- also matches -42.5.
> Perhaps p- ought to match negative integers only?
> 
> `(check-equal? (math-match -42.5 [p+ #f] [p- #t] [_ #f]) #f)`

You are right. I did not notice the buggy + - before.
We need to fix this.

File: racket-cas/racket-cas.rkt (hunk 8 above)
[the remark this answers is not in the document]
Comment (BowenFu) [reply racket-cas/racket-cas.rkt]: You are right. I did not notice the buggy + - before.
We need to fix this.

File: racket-cas/racket-cas.rkt (hunk 9 above)
Comment (BowenFu) [reply racket-cas/racket-cas.rkt]: With math-match*, I'll get an error for the pattern `((TimesTerms (== v) us ...) _)`
`match: incorrect use of ... in pattern in: ...`

File: racket-cas/racket-cas.rkt (hunk 8 above)
[the remark this answers is not in the document]
Comment (BowenFu) [reply racket-cas/racket-cas.rkt]: Is it possible to conjoin the predicates inside find-convention-type?
